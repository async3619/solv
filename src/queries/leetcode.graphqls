type CodeSnippet {
    lang: String!
    langSlug: String!
    code: String!
}

type Question {
    questionId: String!
    questionTitle: String!
    exampleTestcaseList: [String!]!
    metaData: String!
    content: String!
    codeSnippets: [CodeSnippet!]!
}

type Query {
    question(titleSlug: String!): Question
}
