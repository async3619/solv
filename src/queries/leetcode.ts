import gql from "graphql-tag";

export const LeetCodeQuestionDocument = gql`
    query leetCodeQuestion($titleSlug: String!) {
        question(titleSlug: $titleSlug) {
            questionId
            questionTitle
            exampleTestcaseList
            content
            metaData
            codeSnippets {
                lang
                langSlug
                code
            }
        }
    }
`;
