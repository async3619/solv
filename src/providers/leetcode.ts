import { assertParse, is } from "typia";
import { parse as parseDOM } from "node-html-parser";
import { Client, CombinedError, createClient } from "@urql/core";
import { stripHtml } from "string-strip-html";

import { BaseProvider } from "./base";

import { Challenge, InputType } from "../utils/types";
import { LeetCodeQuestionDocument } from "../queries/leetcode";
import type { LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables } from "../queries.data";
import { QuestionMetadata } from "./leetcode.types";

const URL_CHECK_REGEX = /(https?:\/\/)?leetcode\.com\/problems\/([a-z0-9-]+)\/?$/;

export class LeetCodeProvider extends BaseProvider {
    private readonly client: Client;

    public constructor() {
        super(false);

        this.client = createClient({
            url: "https://leetcode.com/graphql",
            requestPolicy: "network-only",
            fetchOptions: () => {
                return {
                    method: "GET",
                    headers: {
                        "content-type": "application/json",
                    },
                };
            },
        });
    }

    public checkUrl(url: string): boolean {
        return URL_CHECK_REGEX.test(url);
    }

    public getId(url: string): string | number {
        const match = URL_CHECK_REGEX.exec(url);
        if (!match) {
            throw new Error("Failed to parse challenge ID.");
        }

        return match[2];
    }

    public beforeExecute(sourceCode: string, input: InputType, challenge: Challenge): string {
        let entryPointName = "solution";
        if (challenge.initialCode) {
            const match = challenge.initialCode.match(/function ([a-zA-Z0-9]+)\(/);
            if (match) {
                entryPointName = match[1];
            }
        }

        return [sourceCode, `console.info(${entryPointName}(...global.input))`].join("\n\n");
    }

    public async retrieve(url: string): Promise<Challenge> {
        const titleSlug = this.getId(url);
        if (!is<string>(titleSlug)) {
            throw new Error("Failed to parse challenge ID.");
        }

        try {
            const { error, data } = await this.client
                .query<LeetCodeQuestionQuery, LeetCodeQuestionQueryVariables>(LeetCodeQuestionDocument, { titleSlug })
                .toPromise();

            if (error) {
                throw error;
            }

            if (!data?.question) {
                throw new Error("Failed to retrieve challenge data.");
            }

            const codeSnippet = data.question.codeSnippets.find(item => item.langSlug === "typescript");
            if (!codeSnippet) {
                throw new Error("This challenge does not support TypeScript.");
            }

            const document = parseDOM(data.question.content);
            const testCases = [...document.querySelectorAll("pre")].filter(item => {
                if (!item.parentNode) {
                    return false;
                }

                if (!item.previousElementSibling) {
                    return false;
                }

                const title = item.previousElementSibling.querySelector("strong.example");
                if (!title) {
                    return false;
                }

                return true;
            });

            const metadata = assertParse<QuestionMetadata>(data.question.metaData);
            const testCaseSummaries = testCases.map(item => stripHtml(item.textContent.trim()).result);
            const inputs: InputType[] = [];
            const outputs: string[] = [];
            for (const summary of testCaseSummaries) {
                const [input, output] = summary.split("\n").map(item => {
                    return item.trim().replace(/^[A-Za-z]*?: /, "");
                });

                const inputParameters = input.split(/, /).map(item => item.split(" = "));
                const data: InputType = [];
                for (const [key, value] of inputParameters) {
                    const index = metadata.params.findIndex(item => item.name === key);
                    if (index === -1) {
                        throw new Error(`Failed to find parameter "${key}".`);
                    }

                    data[index] = value;
                }

                inputs.push(data);
                outputs.push(output.replace("Output: ", ""));
            }

            return {
                id: data.question.questionId,
                title: data.question.questionTitle,
                description: stripHtml(data.question.content).result,
                provider: this,
                initialCode: codeSnippet.code,
                input: inputs,
                output: outputs,
            };
        } catch (e) {
            if (e instanceof CombinedError) {
                if (e.networkError) {
                    throw e.networkError;
                } else if (e.graphQLErrors && e.graphQLErrors.length > 0) {
                    throw e.graphQLErrors[0];
                }
            }

            throw e;
        }
    }
}
