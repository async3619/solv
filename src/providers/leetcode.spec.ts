import { LeetCodeProvider } from "./leetcode";
import { CombinedError } from "@urql/core";

describe("LeetCode Provider", function () {
    let provider: LeetCodeProvider;

    beforeEach(function () {
        provider = new LeetCodeProvider();
    });

    it("should determine valid url", function () {
        expect(provider.checkUrl("https://leetcode.com/problems/valid-parentheses/")).toBeTruthy();
    });

    it("should determine invalid url", function () {
        expect(provider.checkUrl("https://leetcode.com/problems/")).toBeFalsy();
        expect(provider.checkUrl("https://leetcode.com/")).toBeFalsy();
    });

    it("should get id from url", function () {
        expect(provider.getId("https://leetcode.com/problems/valid-parentheses/")).toBe("valid-parentheses");
    });

    it("should throw error when trying to get id from invalid url", function () {
        expect(() => provider.getId("https://leetcode.com/problems/")).toThrow();
    });

    it("should fetch problem information", async function () {
        const problem = await provider.retrieve("https://leetcode.com/problems/valid-parentheses/");

        expect({
            ...problem,
            provider: null,
        }).toMatchSnapshot();
    });

    it("should throw error when trying to fetch problem information from invalid url", async function () {
        await expect(provider.retrieve("https://leetcode.com/problems/")).rejects.toThrow();
    });

    it("should throw error when trying to fetch problem information from invalid problem", async function () {
        await expect(provider.retrieve("https://leetcode.com/problems/invalid-problem/")).rejects.toThrow();
    });

    it("should throw error when network error occurs", async function () {
        Object.defineProperty(provider, "client", {
            // mock urql client
            value: {
                query: () => ({
                    toPromise: () => Promise.reject(new Error("Network error")),
                }),
            },
        });

        await expect(provider.retrieve("https://leetcode.com/problems/valid-parentheses/")).rejects.toThrow();

        Object.defineProperty(provider, "client", {
            // mock urql client
            value: {
                query: () => ({
                    toPromise: () =>
                        Promise.reject(
                            new CombinedError({
                                networkError: new Error("Network error"),
                            }),
                        ),
                }),
            },
        });

        await expect(provider.retrieve("https://leetcode.com/problems/valid-parentheses/")).rejects.toThrow();
    });

    it("should bootstrap source code before executing", async function () {
        const problem = await provider.retrieve("https://leetcode.com/problems/valid-parentheses/");
        const functionName = "thisIsFunctionName";

        problem.initialCode = `function ${functionName}() {}`;

        const sourceCode = provider.beforeExecute("// this is source code", [], problem);
        expect(sourceCode).toBe(
            ["// this is source code", "console.info(thisIsFunctionName(...global.input))"].join("\n\n"),
        );
    });
});
