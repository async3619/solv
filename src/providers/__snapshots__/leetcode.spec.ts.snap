// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeetCode Provider should fetch problem information 1`] = `
{
  "description": "Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.

Example 1:

Input: s = "()"
Output: true

Example 2:

Input: s = "()[]{}"
Output: true

Example 3:

Input: s = "(]"
Output: false

Constraints:

1 4
s consists of parentheses only '()[]{}'.",
  "id": "20",
  "initialCode": "function isValid(s: string): boolean {

};",
  "input": [
    [
      ""()"",
    ],
    [
      ""()[]{}"",
    ],
    [
      ""(]"",
    ],
  ],
  "output": [
    "true",
    "true",
    "false",
  ],
  "provider": null,
  "title": "Valid Parentheses",
}
`;
